### PR TITLE
Add JSON parse and encode function

### LINKED ISSUE #1266: JSON parse
**Is your feature request related to a problem?**
No

**Solution you'd like:**
I whould like the json parser functions from ALIVE to be integrated in to cba. Today a YAML parser exist but one for JSON format whould also be greate. 
Function links below:
- https://github.com/ALiVEOS/ALiVE.OS/blob/4f7ac35a2cc6b257a795eed8d2fbe52e07913c51/addons/x_lib/functions/data/fnc_parseJSON.sqf
- https://github.com/ALiVEOS/ALiVE.OS/blob/4f7ac35a2cc6b257a795eed8d2fbe52e07913c51/addons/x_lib/functions/data/fnc_JSON.sqf
- https://github.com/ALiVEOS/ALiVE.OS/blob/4f7ac35a2cc6b257a795eed8d2fbe52e07913c51/addons/x_lib/functions/data/fnc_encodeJSON.sqf

**Alternatives you've considered:**
Contact maintainers of ALIVE is for rewire and copy paste or rewrite the YAML one. 

**Additional context:**
None

### PR DESCRIPTION
**Pull request by BaerMitUmlaut**
**When merged this pull request will:**
- Add a parse and encode function for JSON

Uses a shift-reduce parser, not too complicated. Handles all valid JSON, but will throw script errors with invalid JSON. Escaped unicode characters are not supported because I didn't want to write a hex to dec function.
Closes #1266.

### CODE REVIEW

**Comment by commy2** [comment]
```sqf
execVM "\x\cba\addons\hashes\test_parseJSON.sqf"
```

<details><summary>Error</summary>

```sqf
15:02:50 "null"
15:02:50 "true"
15:02:50 "1.2"
15:02:50 Error in expression <r = _c;
while { _c = _input deleteAt 0; _c in _numeric } do {
_numStr = _numStr >
15:02:50   Error position: <_c in _numeric } do {
_numStr = _numStr >
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1921
15:02:50 Error in expression <pushBack parseNumber _numStr;
_input = [_c] + _input;
};



case (_c in _consts)>
15:02:50   Error position: <_c] + _input;
};



case (_c in _consts)>
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1925
15:02:50 Error in expression < deleteAt 0;

switch (true) do {

case (_c in _symbols): {
_tokens pushBack _c;
>
15:02:50   Error position: <_c in _symbols): {
_tokens pushBack _c;
>
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1911
15:02:50 """Hello, World!"""
15:02:50 "[]"
15:02:50 "{}"
15:02:50 "[null, true, 1.2, ""Hello, World!""]"
15:02:50 "[{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": []}]}]}]}]}]}]}]"
15:02:50 "null"
15:02:50 "true"
15:02:50 "1.2"
15:02:50 Error in expression <r = _c;
while { _c = _input deleteAt 0; _c in _numeric } do {
_numStr = _numStr >
15:02:50   Error position: <_c in _numeric } do {
_numStr = _numStr >
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1921
15:02:50 Error in expression <pushBack parseNumber _numStr;
_input = [_c] + _input;
};



case (_c in _consts)>
15:02:50   Error position: <_c] + _input;
};



case (_c in _consts)>
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1925
15:02:50 Error in expression < deleteAt 0;

switch (true) do {

case (_c in _symbols): {
_tokens pushBack _c;
>
15:02:50   Error position: <_c in _symbols): {
_tokens pushBack _c;
>
15:02:50   Error Undefined variable in expression: _c
15:02:50 File \x\cba\addons\hashes\fnc_parseJSON.sqf [CBA_fnc_parseJSON]..., line 1911
15:02:50 """Hello, World!"""
15:02:50 "[]"
15:02:50 "{}"
15:02:50 "[null, true, 1.2, ""Hello, World!""]"
15:02:50 "[{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": []}]}]}]}]}]}]}]"
```

</details>

**Comment by commy2** [comment]
After my commit:

```sqf
execVM "\x\cba\addons\hashes\test_parseJSON.sqf"
```

<details><summary>RPT log</summary>

```sqf
15:48:21 [CBA] (hashes) Test OK: (CBA_fnc_parseJSON is defined) x\cba\addons\hashes\test_parseJSON.sqf:14
15:48:21 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:20
15:48:21 [CBA] (hashes) Test OK: (_result isEqualTo _expected) x\cba\addons\hashes\test_parseJSON.sqf:25
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:29
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:33
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:37
15:48:21 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:40
15:48:21 [CBA] (hashes) Test OK: (isNull _result) x\cba\addons\hashes\test_parseJSON.sqf:43
15:48:21 [CBA] (hashes) Test OK: (_result) x\cba\addons\hashes\test_parseJSON.sqf:49
15:48:21 [CBA] (hashes) Test OK: (_result isEqualTo _expected) x\cba\addons\hashes\test_parseJSON.sqf:54
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:58
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:62
15:48:21 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:66
15:48:21 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:69
15:48:21 [CBA] (hashes) Test OK: (isNull _result) x\cba\addons\hashes\test_parseJSON.sqf:72
```

</details>

Not sure what to do with your test.

File: addons/hashes/fnc_parseJSON.sqf [hunk 1]
```diff
@@ -19,6 +19,9 @@ Examples:
         private _json = "{ ""enabled"": true }";
         private _settings = [_json] call CBA_fnc_parseJSON;
         private _enabled = _settings getVariable "enabled";
+
+        loadFile "data\config.json" call CBA_fnc_parseJSON
+        [preprocessFile "data\config.json", true] call CBA_fnc_parseJSON
```
Comment (commy2) [inline addons/hashes/fnc_parseJSON.sqf]: Example for external files.

File: addons/hashes/fnc_encodeJSON.sqf [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    BaerMitUmlaut
+---------------------------------------------------------------------------- */
+SCRIPT(encodeJSON);
+params ["_object"];
+
+if (isNil "_object") exitWith { "null" };
+
+switch (typeName _object) do {
+    case "SCALAR";
+    case "BOOL": {
```
Comment (commy2) [inline addons/hashes/fnc_encodeJSON.sqf]:
```suggestion
    case "CODE";
    case "BOOL": {
```
Curly brackets etc. should not be a problem, but if escaping quote marks is then don't merge this suggestion.

File: addons/hashes/fnc_encodeJSON.sqf [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            }) joinString ", ";
+            "{" + _json + "}"
+        } else {
+            private _json = (_object apply {[_x] call CBA_fnc_encodeJSON}) joinString ", ";
+            "[" + _json + "]"
+        };
+    };
+
+    default {
+        if (isNull _object) exitWith { "null" };
```
Comment (commy2) [inline addons/hashes/fnc_encodeJSON.sqf]: I have this urge to turn objects like the `player` into JSON files.

There should be a check if the _object type supports the `isNull` command.
JSON has no `undefined`, apparently, but not supported types should be treated as `null` with a RPT log message (either ERROR_n or WARNING_n).

```sqf
player call CBA_fnc_encodeJSON; 
```

<details><summary>Error</summary>

```sqf
16:55:23 Error in expression <
"[" + _json + "]"
};
};

default {
if (isNull _object) exitWith { "null" };

pr>
16:55:23   Error position: <isNull _object) exitWith { "null" };

pr>
16:55:23   Error isnull: Type code, expected Object,Group,Script,Config entry,Display (dialog),Control,Network Object,Task,Location
16:55:23 File \x\cba\addons\hashes\fnc_encodeJSON.sqf [CBA_fnc_encodeJSON]..., line 1910
16:55:23 Error in expression <
"[" + _json + "]"
};
};

default {
if (isNull _object) exitWith { "null" };

pr>
16:55:23   Error position: <isNull _object) exitWith { "null" };

pr>
16:55:23   Error Generic error in expression
16:55:23 File \x\cba\addons\hashes\fnc_encodeJSON.sqf [CBA_fnc_encodeJSON]..., line 1910
```

</details>

**Comment by commy2** [comment]
`CBA_fnc_encodeJSON` does not handle HASH types, but idk if that matters.

**Comment by BaerMitUmlaut** [comment]
https://github.com/CBATeam/CBA_A3/blob/d5d3295a7c9f253f8378879cffe2a2e057cbe463/addons/hashes/fnc_encodeJSON.sqf#L63-L71

It handles CBA hashes just fine. The unit test even tests that.

**Comment by commy2** [comment]
You're right. I was using the wrong syntax:
```sqf
private _hash = ["#CBA_HASH#",["a"],[1],nil];
_hash call CBA_fnc_encodeJSON // """#CBA_HASH#"""
[_hash] call CBA_fnc_encodeJSON // "{""a"": 1}"
```

Will you fix the test?

File: addons/hashes/test_parseJSON.sqf [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+_result = [[_data, "phoneNumber"] call CBA_fnc_hashGet select 1, "type"] call CBA_fnc_hashGet;
+_expected = "fax";
+TEST_OP(_result,==,_expected,_fn);
+
+_result = [_data, "newSubscription"] call CBA_fnc_hashGet;
+TEST_FALSE(_result,_fn);
+
+_result = [_data, "companyName"] call CBA_fnc_hashGet;
+TEST_TRUE(isNull _result,_fn);
+
```
Comment (commy2) [inline addons/hashes/test_parseJSON.sqf]:
```suggestion

// ----------------------------------------------------------------------------

_fn = "CBA_fnc_encodeJSON";
LOG("Testing " + _fn);

TEST_DEFINED("CBA_fnc_encodeJSON",_fn);

// ----------------------------------------------------------------------------

```

File: addons/hashes/fnc_encodeJSON.sqf [hunk 5]
```diff
@@ -75,6 +77,10 @@ switch (typeName _object) do {
     };
 
     default {
+        if !(typename _object in (supportinfo "u:allVariables*" apply {_x splitString " " select 1})) exitWith {
```
Comment (commy2) [inline addons/hashes/fnc_encodeJSON.sqf]:
```suggestion
        if !(typeName _object in (supportInfo "u:allVariables*" apply {_x splitString " " select 1})) exitWith {
```

File: addons/hashes/fnc_encodeJSON.sqf [hunk 6]
```diff
@@ -48,15 +49,16 @@ switch (typeName _object) do {
...
+            ["""", "\"""],
             [toString [8], "\b"],
             [toString [12], "\f"],
             [endl, "\n"],
+            [toString [10], "\n"],
             [toString [13], "\r"],
             [toString [9], "\t"]
         ];
-        str _object
+        """" + _object + """"
```
Comment (commy2) [inline addons/hashes/fnc_encodeJSON.sqf]: Uh, how is this different now? I don't get it.

File: addons/hashes/test_parseJSON.sqf [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+{
+    private _useHashes = _x;
+
+    {
+        diag_log _x;
+        private _input  = _x;
+        private _object = [_x, _useHashes] call CBA_fnc_parseJSON;
+        private _output = [_object] call CBA_fnc_encodeJSON;
+        TEST_OP(_input,==,_output,_fn);
```
Comment (commy2) [inline addons/hashes/test_parseJSON.sqf]: `TEST_OP` is logging asynchronous, `diag_log` at command execution. This makes the logging appear in RPT out of order.

Either use `LOG_1(_x)` (requires `#define DEBUG_MODE_FULL` in L1), or `#define DEBUG_SYNCHRONOUS` in L1.

File: addons/hashes/fnc_encodeJSON.sqf (hunk 6 above)
Comment (BaerMitUmlaut) [reply addons/hashes/fnc_encodeJSON.sqf]: `str` escapes quote signs, which we already do with backslashes.

grafik

Comment (commy2) [reply addons/hashes/fnc_encodeJSON.sqf]: Wow, that's fucked.

Comment (commy2) [reply addons/hashes/fnc_encodeJSON.sqf]: Warrants a comment.
```suggestion
        // stringify without escaping inter string quote marks
        """" + _object + """"
```

**Comment by commy2** [comment]
```sqf
execVM "\x\cba\addons\hashes\test_parseJSON.sqf"
```

<details><summary>RPT log</summary>

```sqf
10:42:58 [CBA] (hashes) Test OK: (CBA_fnc_parseJSON is defined) x\cba\addons\hashes\test_parseJSON.sqf:14
10:42:58 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:20
10:42:58 [CBA] (hashes) Test OK: (_result isEqualTo _expected) x\cba\addons\hashes\test_parseJSON.sqf:25
10:42:58 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:29
10:42:58 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:33
10:42:58 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:37
10:42:58 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:41
10:42:58 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:45
10:42:58 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:48
10:42:58 [CBA] (hashes) Test OK: (isNull _result) x\cba\addons\hashes\test_parseJSON.sqf:51
10:42:59 [CBA] (hashes) Test OK: (_result) x\cba\addons\hashes\test_parseJSON.sqf:57
10:42:59 [CBA] (hashes) Test OK: (_result isEqualTo _expected) x\cba\addons\hashes\test_parseJSON.sqf:62
10:42:59 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:66
10:42:59 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:70
10:42:59 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:74
10:42:59 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:78
10:42:59 [CBA] (hashes) Test OK: (_result == _expected) x\cba\addons\hashes\test_parseJSON.sqf:82
10:42:59 [CBA] (hashes) Test OK: (not (_result)) x\cba\addons\hashes\test_parseJSON.sqf:85
10:42:59 [CBA] (hashes) Test OK: (isNull _result) x\cba\addons\hashes\test_parseJSON.sqf:88
10:42:59 [CBA] (hashes) Test OK: (CBA_fnc_encodeJSON is defined) x\cba\addons\hashes\test_parseJSON.sqf:95
10:42:59 "null"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "true"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "1.2"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 """Hello, World!"""
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "{}"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[null, true, 1.2, ""Hello, World!""]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": []}]}]}]}]}]}]}]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "null"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "true"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "1.2"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 """Hello, World!"""
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "{}"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[null, true, 1.2, ""Hello, World!""]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 "[{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": [{""nested"": []}]}]}]}]}]}]}]"
10:42:59 [CBA] (hashes) Test OK: (_input == _output) x\cba\addons\hashes\test_parseJSON.sqf:118
10:42:59 [CBA] (hashes) Test OK: (count _properties == 6) x\cba\addons\hashes\test_parseJSON.sqf:126
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
10:42:59 [CBA] (hashes) Test OK: (typeName _value == _x) x\cba\addons\hashes\test_parseJSON.sqf:129
```

</details>

**Pull request merged**